include/deal.II/physics/notation.templates.h

```diff
@@ -0,0 +1,763 @@
+    {
+      FullMatrix<Number> out (Tensor<2,dim,Number>::n_independent_components,
+                              Tensor<2,dim,Number>::n_independent_components);
+      const unsigned int n_rows = out.m();
+      const unsigned int n_cols = out.n();
+      for (unsigned int r=0; r<n_rows; ++r)
+        {
+          const std::pair<unsigned int,unsigned int> indices_ij = indices_from_component(r,false);
+          Assert(indices_ij.first < dim, ExcInternalError());
+          Assert(indices_ij.second < dim, ExcInternalError());
+          const unsigned int &i = indices_ij.first;
+          const unsigned int &j = indices_ij.second;
+
+          for (unsigned int c=0; c<n_cols; ++c)
+            {
+              const std::pair<unsigned int,unsigned int> indices_kl = indices_from_component(c,false);
+              Assert(indices_kl.first < dim, ExcInternalError());
+              Assert(indices_kl.second < dim, ExcInternalError());
+              const unsigned int &k = indices_kl.first;
+              const unsigned int &l = indices_kl.second;
+
+              out(r,c) = t[i][j][k][l];
+            }
+        }
+      return out;
+    }
+
+
+    template<int dim>
+    template<typename Number>
+    FullMatrix<Number>
+    Kelvin<dim>::unroll_to_matrix (const SymmetricTensor<4,dim,Number> &st)
+    {
+      FullMatrix<Number> out (SymmetricTensor<2,dim,Number>::n_independent_components,
+                              SymmetricTensor<2,dim,Number>::n_independent_components);
+      const unsigned int n_rows = out.m();
+      const unsigned int n_cols = out.n();
+      for (unsigned int r=0; r<n_rows; ++r)
+        {
+          const std::pair<unsigned int,unsigned int> indices_ij = indices_from_component(r,true);
+          Assert(indices_ij.first < dim, ExcInternalError());
+          Assert(indices_ij.second < dim, ExcInternalError());
+          Assert(indices_ij.second >= indices_ij.first, ExcInternalError());
+          const unsigned int &i = indices_ij.first;
+          const unsigned int &j = indices_ij.second;
+
+          for (unsigned int c=0; c<n_cols; ++c)
+            {
+              const std::pair<unsigned int,unsigned int> indices_kl = indices_from_component(c,true);
+              Assert(indices_kl.first < dim, ExcInternalError());
+              Assert(indices_kl.second < dim, ExcInternalError());
+              Assert(indices_kl.second >= indices_kl.first, ExcInternalError());
+              const unsigned int &k = indices_kl.first;
+              const unsigned int &l = indices_kl.second;
+
+              const double factor = matrix_component_factor(r,c,true);
+
+              out(r,c) = factor*st[i][j][k][l];
+            }
+        }
+      return out;
+    }
+
+
+    template<int dim>
+    template<typename Number>
+    void
+    Kelvin<dim>::roll_to_tensor (Number &s,
+                                 const Vector<Number> &vec)
+    {
+      Assert(vec.size() == 1, ExcMessage("Input vector has incorrect size."));
```

Comment:
I think `ExcDimensionMismatch()` is a more descriptive as it gives you actual numbers.

Reply:
same below.